### PR TITLE
Initial draft for review

### PR DESCRIPTION
**Pull request by outsidecoder**
### What

Create Unit test examples and some documentation.  Documentation needs further work, including creating slides, which I've started on Trello.

### How to review

Read and check the tests work okay and are useful.

### Who can review

Anyone

### CODE REVIEW

File: training/unitTesting/goLang/3.dependencyInjection/service/service_tes.g [hunk 1]
```diff
@@ -0,0 +1,104 @@
+package service
```
Comment (gedge) [inline training/unitTesting/goLang/3.dependencyInjection/service/service_tes.g:1]: delete this file?

File: guides/NEW_LIBRARY.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+* A login for the concourse CI environment
+
+* [dp-cli](https://github.com/ONSdigital/dp-cli)
+
+* Git:
+
+  `$ brew install git`
+
+* Fly CLI (can be downloaded from the links in the corner of the concourse UI)
```
Comment (nshumoogum) [inline guides/NEW_LIBRARY.md:23]: Possibly link to concourse UI here:

".... links in the bottom right corner of the [concourse UI](https://concourse.onsdigital.co.uk/)"

File: training/unit-testing/go-lang/README.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
+Unit Testing in Go
+==================
+
+### Pre reading & Pre requisites
+It's advised that you are comfortable with GoLang or another programming or scripting language.  The setup is focussed on linux based operating systems..
```
Comment (nshumoogum) [inline training/unit-testing/go-lang/README.md:5]: Should we rename GoLang to Go if not at least Golang (not sure why the L is capitalised here).

fyi - "The language is often referred to as Golang because of its domain name, golang.org , but the proper name is Go."

File: training/unit-testing/go-lang/README.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+A useful approach, particularly for potentially difficult to test code, is Test Driven Development, which involves writing the test at the same time or before the code you need to test. So effectively designing easy to test code from the start.
+- [Golang Test Driven Development walk through](https://itnext.io/how-to-tdd-a-console-application-to-achieve-100-coverage-starting-from-main-test-go-934a617b080f) 
+
+### Recommended Go Tutorials
+- [GoConvey Tutorial 1](https://www.smartystreets.com/blog/2015/02/go-testing-part-1-vanillla/)
+- [GoConvey Tutorial 2](https://www.smartystreets.com/blog/2015/02/go-testing-part-2-running-tests/)
+- [GoConvey Tutorial 3](https://www.smartystreets.com/blog/2015/05/go-testing-part-3-convey-behavior/)
+
+### Code Coverage
+When we test certain paths of execution are tested, but often not all.  Knowing which paths are tested is useful, but often the same paths need to be tested multiple time, so 100% coverage doesn't mean 100% tested.
```
Comment (nshumoogum) [inline training/unit-testing/go-lang/README.md:89]: Extra whitespace between `often not all. ` and ` Knowing which paths`

File: training/unit-testing/go-lang/README.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+- [Excellent Golang code coverage walk-through](https://blog.golang.org/cover)
+- [Useful time saving suggestion on stackoverflow](https://stackoverflow.com/a/27284510)
+
+### Assessing How Fast Your Code Runs
+Systems often need to run efficiently, such as a response from a website.  Go incorporates a benchmarking routines, which will time how long the code will take to run.  It clever though as it repeats the tests until it feel it's got an accurate time.
```
Comment (nshumoogum) [inline training/unit-testing/go-lang/README.md:94]: Again extra whitespace between `a website. ` and ` Go incorporates` and then `run. ` and ` It clever` - Im now wondering if this wasn't by accident?

File: training/unit-testing/go-lang/README.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+### Other Interesting Testing Related Information
+- [Dot import for Convey](https://smartystreets.com/blog/2015/05/go-testing-part-3-convey-behavior/)
+
+##Wordier Information
+###Background To Unit Testing
+Unit testing is the process of testing a portion of an application or service to ensure it meets our requirements.  The portion of code is ideally a single function, but is typically less than a few go files.
+
+Ideally we should only test the portion of the application we are focused on, and shouldn't connect with other areas of a system.  This can be difficult, and so sometimes different portions of the system are 'mocked'.
+
+### Is Unit Testing Ever Pointless?
```
Comment (nshumoogum) [inline training/unit-testing/go-lang/README.md:120]: I wonder if there should be a piece here regarding duplicating tests across functions that wrap one another.

For example a handle method that then calls off to a validate method on an object should only contain 1 or 2 tests to cover for success and failure coming from the validate method even if the validate method contains 8 individual pathways which could lead to failure.

Those 8 individual pathways to failure should be unit tested against the validate method to get close to 100% test coverage.

Side note: if your method or function has so many unit tests against it, there is an argument it should be refactored to reduce complexity, once refactored we can apply the logic in the above paragraph and this should result in clearer unit tests for readability and better coverage.

File: training/unit-testing/go-lang/benchmarking/benchmark_test.go [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+	. "github.com/smartystreets/goconvey/convey"
+)
+
+//init is used here to ensure current goroutine exclusively tied to current OS thread
+func init() {
+	runtime.LockOSThread()
+}
+
+// TestMain demonstrates the use of TestMain and also attaching the tests to the main OS thread.
+// Can be run with th following command `go test -run=XXX -bench=.`This may be usedful for some Go functionality that needs to have the test thread attached to the OS thread.
```
Comment (nshumoogum) [inline training/unit-testing/go-lang/benchmarking/benchmark_test.go:18]: Typo `th` instead of `the`

File: training/unit-testing/go-lang/helloworld/hello_world_test.go [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+	})
+}
+
+// TestHelloWorld3 is a half finished test where we want Convey to only test the tests that are finished.
+// Focus Convey - all sub Convey's need to be FocusConvey's otherwise they are ignored.
+// SkipConvey - all Convey's and tests underneath will be skipped
+// SkipSo - this test is skipped
+func TestHelloWorld3(t *testing.T) {
+	const MaxDifference = 3
+	FocusConvey("Given two numbers check x is greater than y", t, func() { //ONLY TESTS WITH FOCUS ARE RUN.
```
Comment (nshumoogum) [inline training/unit-testing/go-lang/helloworld/hello_world_test.go:79]: I didn't know about this function, you learn something new everyday 😄

**Review by nshumoogum — CHANGES_REQUESTED** [review]

File: training/unit-testing/go-lang/README.md (hunk 4 above)
Comment (outsidecoder) [reply training/unit-testing/go-lang/README.md:89]: Two spaces after a full stop is what I was taught for accessibility/readability.  Is one space after full stop ONS house style?

File: training/unit-testing/go-lang/README.md (hunk 5 above)
Comment (outsidecoder) [reply training/unit-testing/go-lang/README.md:94]: See comment above.

File: training/unit-testing/go-lang/README.md (hunk 6 above)
Comment (outsidecoder) [reply training/unit-testing/go-lang/README.md:120]: Yip that sounds really good.  Do you happen to have an example I can point to?  Even a before and after?

File: training/unit-testing/go-lang/benchmarking/benchmark_test.go (hunk 7 above)
Comment (outsidecoder) [reply training/unit-testing/go-lang/benchmarking/benchmark_test.go:18]: Thanks - fixed.

File: training/unit-testing/go-lang/helloworld/hello_world_test.go (hunk 8 above)
Comment (outsidecoder) [reply training/unit-testing/go-lang/helloworld/hello_world_test.go:79]: 👌

File: training/unit-testing/go-lang/README.md (hunk 6 above)
Comment (nshumoogum) [reply training/unit-testing/go-lang/README.md:120]: As per direct message the dp-identity-api has been written in a way where unit test coverage is not duplicated across handlers and bespoke methods/functions. Check out the Tokens handler and the ValidateCredentials method with their respective unit tests.

File: training/unit-testing/go-lang/README.md (hunk 4 above)
Comment (nshumoogum) [reply training/unit-testing/go-lang/README.md]: I did not know that but I can see that being a thing - no need to change this and happy for you to point this out to me on my docs :)

File: training/unit-testing/go-lang/README.md [hunk 9]
```diff
@@ -49,75 +31,81 @@ In your terminal:
...
 
 #### Order of Execution
-GoConvey's execution path is a little quirky so have a look at the example below and the resources.
+GoConvey's execution path is a little quirky so have a look at the example below and the linked resources.
 - [Downloadable order of execution example](#orderexecution/execution_order_example_test.go)
 - [GoConvey explanation](https://github.com/smartystreets/goconvey/wiki/Execution-order)
 
 #### Assertions
-- When comparing objects in your tests you can use [GoConvey's Assertions](/assertions/README.md). 
+When comparing objects in your tests you can use [GoConvey's Assertions](/assertions/README.md), however when comparing complex `structs` such as `errors` can be troublesome.
```
Comment (nshumoogum) [inline training/unit-testing/go-lang/README.md:47]: errors are a built-in type but under the hood represent an interface to allow users to create their own errors - this can make it harder to test as you may want to check the stringified error value or if you have made a custom error which contains other functions or values you would need to caste the error to your custom error.

Either way this is easy, average or difficult depending on your knowledge of go.

Basically I would not put an opinion of the difficulty in the documentation as it is subjective

Comment (nshumoogum) [reply training/unit-testing/go-lang/README.md:47]: Instead maybe rephrase this that we can use DeepEqual to assist with comparing structs

File: training/unit-testing/go-lang/README.md [hunk 10]
```diff
@@ -49,75 +31,81 @@ In your terminal:
...
+When we test, certain paths of execution are run, but often not all.  Code Coverage tells us which code was executed during a test.  Sometimes the same paths need to be tested multiple times, but its good to limit this.  It's also important to note that 100% code coverage doesn't mean 100% tested.
+
+Here are two useful links to read:
 - [Excellent Golang code coverage walk-through](https://blog.golang.org/cover)
 - [Useful time saving suggestion on stackoverflow](https://stackoverflow.com/a/27284510)
 
-### Assessing How Fast Your Code Runs
-Systems often need to run efficiently, such as a response from a website.  Go incorporates a benchmarking routines, which will time how long the code will take to run.  It clever though as it repeats the tests until it feel it's got an accurate time.  
-- [Downloadable benchmarking example](benchmarking/benchmark_test.go)
+[dp-identify-api]() is a great ONS example of a nicely designed, loosely coupled repo.  Its design also limits the code which is re-run as part of the tests, therefore saving resources.
```
Comment (nshumoogum) [inline training/unit-testing/go-lang/README.md:81]: Missing link to repository - https://github.com/ONSdigital/dp-identity-api

**Review by nshumoogum — CHANGES_REQUESTED** [review]

File: training/unit-testing/go-lang/README.md (hunk 9 above)
Comment (outsidecoder) [reply training/unit-testing/go-lang/README.md:47]: Yip that makes sense.  Thanks Nathan.

File: training/unit-testing/go-lang/README.md (hunk 10 above)
Comment (outsidecoder) [reply training/unit-testing/go-lang/README.md:81]: Whoops - thanks!

**Comment by nshumoogum** [comment]
Closing as new PR here https://github.com/ONSdigital/dp/pull/208
